lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py

```diff
@@ -0,0 +1,268 @@
+    boto3_conn,
+    ec2_argument_spec,
+    get_aws_connection_info,
+    AWSRetry,
+    HAS_BOTO3,
+    camel_dict_to_snake_dict
+)
+
+try:
+    from botocore.exceptions import ClientError, NoCredentialsError
+except ImportError:
+    pass  # will be picked up by imported HAS_BOTO3
+
+
+def check_eigw_status(module, client, eigw_id, wait, timeout, action):
+    """
+    Check EIGW status.
+
+    module  : AnsibleModule object
+    client  : boto3 client connection object
+    eigw_id : ID of the EIGW to check status
+    wait    : Whether to wait or not
+    timeout : Timeout in seconds to wait for status
+    action  : The state action we are performing attach or detach
+    """
+    if action == 'attach':
+        desired_state = 'attached'
+        intermediate_state = 'attaching'
+    elif action == 'detach':
+        desired_state = 'detached'
+        intermediate_state = 'detaching'
+    else:
+        module.fail_json(msg='Bad attachment action provided to function check_eigw_status(): {0}.'.format(action))
+
+    pause = 2
+    backoff = 2
+    timer = time.time() + timeout
+    while time.time() < timer:
+        time.sleep(pause)
+        try:
+            check_resp = client.describe_egress_only_internet_gateways(EgressOnlyInternetGatewayIds=eigw_id)
+        except ClientError as e:
+            module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+
+        # We have achieved desired state so we return success
+        if check_resp['EgressOnlyInternetGateways'][0]['Attachments'][0]['State'] == desired_state:
+            return True, 'success'
+        # Still in intermediate state so we continue to wait
+        elif check_resp['EgressOnlyInternetGateways'][0]['Attachments'][0]['State'] == intermediate_state:
+            pause = pause * backoff
+            continue
+        # Attachment status is not desired or intermediate state so likely something went wrong.
+        else:
+            return False, 'bad attachment state'
+
+    # Loop finished without success so we return failure
+    return False, 'timeout'
+
+
+@AWSRetry.backoff()
+def delete_eigw(module, client, eigw_id):
+    """
+    Delete EIGW.
+
+    module     : AnsibleModule object
+    client     : boto3 client connection object
+    eigw_id    : ID of the EIGW to delete
+    """
+    try:
+        response = client.delete_egress_only_internet_gateway(DryRun=module.check_mode, EgressOnlyInternetGatewayId=eigw_id)
+    except ClientError as e:
+        # When boto3 method is run with DryRun=True it returns an error on success
+        # We need to catch the error and return something valid
+        if e.response.get('Error').get('Code') == "DryRunOperation":
+            return {'changed': True, 'gateway_id': None}
+        else:
+            module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+
+    return {'changed': response['ReturnCode']}
```

Comment:
Wouldn't this end up returning `changed=200` to callers? Usually we keep this a boolean.

Reply:
I had to review the boto3 docs to be sure but the delete_egress_only_internet_gateway function response syntax is as follows:
```
{
    'ReturnCode': True|False
}
```

http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Client.delete_egress_only_internet_gateway

So its returning a bool as ReturnCode. Seems counter-intuitive of the AWS API.